### PR TITLE
Bump minimum Python to 3.10

### PR DESCRIPTION
**Pull request by nstarman**
v6.1 will be py3.10+.

- [x] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: .github/workflows/ci_cron_weekly.yml
Comment (nstarman) [inline .github/workflows/ci_cron_weekly.yml:95]: I don't think ubuntu-20.04 supports py3.12.
@pllim can we change this to ubuntu-latest?

Comment (nstarman) [reply .github/workflows/ci_cron_weekly.yml:95]: Or Ubuntu-22.04?

Comment (nstarman) [reply .github/workflows/ci_cron_weekly.yml]: I tried bumping

Comment (nstarman) [reply .github/workflows/ci_cron_weekly.yml]: Hm. That's not the reason. Something about Docker.

Comment (nstarman) [reply .github/workflows/ci_cron_weekly.yml]: I can reduce it back to ubuntu-20.04, unless we should just keep it on ubuntu-latest?

**Comment by nstarman** [comment]
I bumped SciPy to 1.6, but looking at PyPI, SciPy v1.7.2 is the minimum version officially supporting py3.10 (https://pypi.org/project/scipy/1.7.2/). I guess we need to bump to SciPy v1.7.2?

File: .circleci/config.yml
Comment (nstarman) [inline .circleci/config.yml:89]: Need to regenerate the comparison images.

File: .ruff.toml
Comment (eerovaher) [inline .ruff.toml:223]: Why is this being added?

Comment (nstarman) [reply .ruff.toml:223]: It only turns on with py3.10+, so we didn't detect it before.

Comment (nstarman) [reply .ruff.toml:223]: I put it on the temporary list because I expect that future versions of python will fix the performance aspect of this syntax, doing some kind of in-lining. When the Union operation is as fast as a tuple then we should let Ruff fix it — `isinstance(x, this | that)` is semantically better.

Comment (neutrinoceros) [reply .ruff.toml:223]: Makes sense to me. I'd suggest noting the reason in the comment 
```suggestion
    # pyupgrade (UP)
    "UP038",  # isinstance using union separators. The code is slower as of Python 3.10-3.12

```

File: CONTRIBUTING.md
Comment (eerovaher) [inline CONTRIBUTING.md:124]: If you want the Python version written here to be one that is actually supported by `astropy` then I recommend you write the latest supported version (i.e. 3.12) here. That way this page stays valid for longer.

File: CONTRIBUTING.md
Comment (eerovaher) [inline CONTRIBUTING.md:126]: This is badly out of date. Perhaps it's best to replace an explicit listing of dependencies with a reference to `pyproject.toml`.

File: astropy/cosmology/funcs/comparison.py
Comment (eerovaher) [inline astropy/cosmology/funcs/comparison.py:12]: [This change does make the file comply with PEP 8](https://peps.python.org/pep-0008/#module-level-dunder-names), but it's not related to dropping support for Python 3.9, so in this pull request it's just adding clutter.

File: astropy/coordinates/tests/test_frames_with_velocity.py
Comment (eerovaher) [inline astropy/coordinates/tests/test_frames_with_velocity.py:65]: I suspect there might be a few more places where using the dictionary merge operator could lead to fewer lines of code that are not marked with a easily greppable comment, but looking for them would be beyond the scope of this pull request and there's no urgent need to do that anyways.

File: CONTRIBUTING.md
Comment (nstarman) [reply CONTRIBUTING.md:124]: Or the more generic "compatible with our supported versions of Python (see pyproject.toml)"

File: astropy/coordinates/tests/test_frames_with_velocity.py
Comment (nstarman) [reply astropy/coordinates/tests/test_frames_with_velocity.py:65]: I wish pyupgrade would catch these. Feels deterministic.

File: CONTRIBUTING.md
Comment (nstarman) [reply CONTRIBUTING.md]: ✅

File: astropy/cosmology/funcs/comparison.py
Comment (nstarman) [reply astropy/cosmology/funcs/comparison.py]: ✅

File: CONTRIBUTING.md
Comment (nstarman) [reply CONTRIBUTING.md]: ✅

**Comment by nstarman** [comment]
Documentation link check is failing because Astropy v6.0 isn't out yet.

**Comment by eerovaher** [comment]
If it is possible to update the minimum required `numpy` and `scipy` versions without having to update the reference images then it would be a good idea to open a separate pull request for that.

**Comment by saimn** [comment]
3.10 is supposed to be supported until April, so if we want to have the option to release a minor version before that (e.g. for Numpy 2.0 in February) this PR  should wait.

File: .github/workflows/ci_cron_weekly.yml
Comment (nstarman) [reply .github/workflows/ci_cron_weekly.yml]: Latest keeps it properly updated. SGTM.

**Comment by neutrinoceros** [comment]
@nstarman This was just merged and could be cleaned in this PR https://github.com/astropy/astropy/blob/b259f3f1978e565776049ac64cc2445f78a11c37/tox.ini#L85-L87

**Comment by eerovaher** [comment]
The planned Numpy 2.0 release date is not in February anymore like it was when https://github.com/astropy/astropy/pull/15603#issuecomment-1861612870 was posted. Are we still considering making an `astropy` release before April? If not then it would be good to remove Python 3.9 support sooner rather than later.

**Comment by pllim** [comment]
> Are we still considering making an astropy release before April?

I don't think so. Non bugfix release is every 6 months, so feature freeze isn't until April. See https://github.com/astropy/astropy/wiki/Release-Calendar

**Comment by WilliamJamieson** [comment]
I am in favor of merging this sooner rather than later if there are no plans to release Astropy before April 5.

File: .circleci/config.yml
Comment (pllim) [reply .circleci/config.yml:89]: You can do that as part of this PR by updating the hash.

**Comment by pllim** [comment]
I can review but I want to see green CI first, and maybe rebase again to make sure we pick up the latest changes after that.

File: .circleci/config.yml
Comment (nstarman) [reply .circleci/config.yml:89]: Where are these hashes stored?

Comment (pllim) [reply .circleci/config.yml:89]: It is mentioned somewhere, maybe https://docs.astropy.org/en/latest/development/testguide.html#image-tests-with-pytest-mpl

@neutrinoceros recently did it so he would know best.

Comment (neutrinoceros) [reply .circleci/config.yml:89]: that would be `astropy/tests/figures/` !

**Comment by nstarman** [comment]
Some numpy dev failure, may not be specific to this PR.

**Comment by pllim** [comment]
I triggered the equivalent job on current `main` to make sure this failure is unrelated: https://github.com/astropy/astropy/actions/runs/8025362924/job/21925825752

**Comment by pllim** [comment]
OK it is indeed unrelated:

* https://github.com/astropy/astropy/issues/16100

**Comment by pllim** [comment]
Please rebase to pick up numpy-dev fixes and resolve conflict. Thanks!

File: .github/workflows/ci_cron_weekly.yml
Comment (pllim) [inline .github/workflows/ci_cron_weekly.yml:104]: Did this cause the exotic arch job failure to start?

**Review by pllim — APPROVED** [review]
I like the clean-up but I won't merge yet to give other maintainers a chance to review. Though if you don't hear back from anyone else, we should get this in sooner than later to give downstream a chance to test this via nightly wheel.

Please ping me next week if no one else approves.

Thanks!

**Comment by pllim** [comment]
p.s. Auto merge won't work because the branch protection rules need updating but I cannot do that until after merge. I can override the protection when we have enough approvals.

File: docs/development/workflow/git_edit_workflow_examples.rst
Comment (mhvk) [inline docs/development/workflow/git_edit_workflow_examples.rst:62]: If there are any other changes requested, might as well change this to 3.12 so we don't have to repeat it for a while...

File: docs/units/type_hints.rst
Comment (mhvk) [inline docs/units/type_hints.rst:68]: Do you want to change this one to `Quantity[u.m] | None` too?

**Review by mhvk — APPROVED** [review]
Great, thanks! Good as is, but if you address the second comment, might as well do the first too.

File: docs/units/type_hints.rst
Comment (nstarman) [reply docs/units/type_hints.rst]: ✅

File: docs/development/workflow/git_edit_workflow_examples.rst
Comment (nstarman) [reply docs/development/workflow/git_edit_workflow_examples.rst]: ✅

File: astropy/units/tests/test_quantity_annotations.py
Comment (eerovaher) [inline astropy/units/tests/test_quantity_annotations.py:6]: I'd get rid of the `# STDLIB` and `# THIRD PARTY` comments.

**Review by eerovaher — COMMENTED** [review]
I have just one small remark.

File: astropy/units/tests/test_quantity_annotations.py
Comment (nstarman) [reply astropy/units/tests/test_quantity_annotations.py]: Done.

**Review by eerovaher — APPROVED** [review]
The patch here looks good to me. The commit history is a bit messy, but that could be fixed by squashing.

**Comment by nstarman** [comment]
> Please ping me next week if no one else approves.

@pllim.

**Comment by pllim** [comment]
Yup yup... 😬

**Comment by pllim** [comment]
Hmm... why did codecov fail? Lemme restart it when I can and see...

https://github.com/astropy/astropy/actions/runs/8208654788/job/22452519783

**Pull request merged**
